modules/siddhi-extensions/input-transports/jms-input-transport/src/main/java/org/wso2/siddhi/extension/input/transport/jms/util/JMSListener.java

```diff
@@ -0,0 +1,192 @@
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
+package org.wso2.siddhi.extension.input.transport.jms.util;
+
+import org.apache.log4j.Logger;
+
+import javax.jms.Connection;
+import javax.jms.ConnectionFactory;
+import javax.jms.JMSException;
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+import java.util.Hashtable;
+
+/**
+ * The revamped JMS Transport listener implementation. Creates {@link JMSTaskManager} instances
+ * for each service requesting exposure over JMS, and stops these if they are undeployed / stopped.
+ * <p/>
+ * A service indicates a JMS Connection factory definition by name, which would be defined in the
+ * JMSListner on the axis2.xml, and this provides a way to reuse common configuration between
+ * services, as well as to optimize resources utilized
+ * <p/>
+ * If the connection factory name was not specified, it will default to the one named "default"
+ * {@see JMSConstants.DEFAULT_CONFAC_NAME}
+ * <p/>
+ * If a destination JNDI name is not specified, a service will expect to use a Queue with the same
+ * JNDI name as of the service. Additional Parameters allows one to bind to a Topic or specify
+ * many more detailed control options. See package documentation for more details
+ * <p/>
+ * All Destinations / JMS Administered objects used MUST be pre-created or already available
+ */
+public class JMSListener implements Runnable {
+
+    private static final Logger log = Logger.getLogger(JMSListener.class);
+    private final String listenerName;
+    private final JMSTaskManager jmsTaskManager;
+
+    public JMSListener(String listenerName, JMSTaskManager jmsTaskManager) {
+        this.listenerName = listenerName;
+        this.jmsTaskManager = jmsTaskManager;
+    }
+
+
+    /**
+     * Listen for JMS messages on behalf of the given service
+     */
+    private void start() {
+
+        JMSTaskManager stm = jmsTaskManager;
+        boolean connected = false;
+
+        /* The following parameters are used when trying to connect with the JMS provider at the start up*/
+        int r = 1;
+        long retryDuration = 10000;
+        double reconnectionProgressionFactor = 2.0;
+        long maxReconnectDuration = 1000 * 60 * 60; // 1 hour
+
+
+        // First we will check whether jms provider is started or not, as if not it will throw a continuous error log
+        // If jms provider not started we will wait for exponentially increasing time intervals,
+        // till the provider is started
+        while (!connected) {
+            boolean jmsProviderStarted = checkJMSConnection(stm);
+            if (jmsProviderStarted) {
+                log.info("Connection attempt: " + r + " for JMS Provider for listener: " + listenerName
+                        + " was successful!");
+                connected = true;
+                stm.start();
+
+                for (int i = 0; i < 3; i++) {
```

Comment:
It doesn't make much sense to loop thrice to me. What do you exactly intend to do here? If you're waiting for a consumer to start, isn't there an API to check whether it is started?